### PR TITLE
Refactor shapes

### PR DESCRIPTION
**Pull request by inducer**
This pushes https://github.com/inducer/modepy/pull/16 towards relying even more heavily on the shapes stuff.

- [ ] Test unmodified meshmode against this: https://github.com/inducer/meshmode/pull/95
- [ ] Looks like CI isn't running here. PR for CI (same branch!): https://github.com/inducer/modepy/pull/19

### CODE REVIEW

File: modepy/matrices.py
Comment (alexfikl) [inline modepy/matrices.py]:
```suggestion
        like those returned by :func:`modepy.Basis.functions`.
```
? Since `orthonormal_basis_for_shape` seems to return a `Basis` class.

File: modepy/modes.py
Comment (alexfikl) [inline modepy/modes.py]:
```suggestion
    from modepy.shapes import node_tuples_for_shape
    mode_ids = node_tuples_for_shape(Simplex(dims), n)
```

File: modepy/modes.py
Comment (alexfikl) [inline modepy/modes.py:713]:
```suggestion
            "Use basis_for_shape instead. "
```
Same as for `tensor_product_basis`.

File: modepy/modes.py
Comment (alexfikl) [inline modepy/modes.py:691]:
```suggestion
            "Use basis_for_shape instead. "
```
? or make `TensorProductBasis` public?

File: modepy/nodes.py
Comment (alexfikl) [inline modepy/nodes.py]: Don't feel strongly about it at all, but `biunit_nodes_for_shape`?

File: modepy/nodes.py
Comment (alexfikl) [inline modepy/nodes.py]: Useless `mp` import.

File: modepy/shapes.py
Comment (alexfikl) [inline modepy/shapes.py]:
```suggestion
    :returns: an :class:`~numpy.ndarray` of shape `(dim, nvertices)`.
```
? Since it's just going to show `ndarray`?

File: modepy/shapes.py
Comment (alexfikl) [inline modepy/shapes.py]: Doesn't seem to inherit from `Shape`?

File: modepy/shapes.py
Comment (alexfikl) [inline modepy/shapes.py]:
```suggestion

        The volume :class:`Shape` from which this face descends.
```

File: modepy/shapes.py
Comment (alexfikl) [inline modepy/shapes.py:260]:
```suggestion

        The face index in :attr:`volume_shape` of this face.
```

File: modepy/shapes.py
Comment (alexfikl) [inline modepy/shapes.py]:
```suggestion

        A tuple of indices into the vertices returned by
        :func:`biunit_vertices_for_shape` for the :attr:`volume_shape`.
```

File: modepy/shapes.py
Comment (alexfikl) [inline modepy/shapes.py]:
```suggestion
    .. attribute:: map_to_volume
    
        A :class:`~collections.abc.Callable` that takes an array of
        size `(dim, nnodes)` of unit nodes on the face represented by
        *face_vertices* and maps them to the :attr:`volume_shape`.
```

File: modepy/shapes.py
Comment (alexfikl) [inline modepy/shapes.py]:
```suggestion
                volume_vertex_indices=tuple(fvi),
```
?

File: modepy/tools.py
Comment (alexfikl) [inline modepy/tools.py]: This is a bit weird in this section. New section?

File: test/test_nodes.py
Comment (alexfikl) [inline test/test_nodes.py:40]: Why the weird randon number generator?

**Review by alexfikl — APPROVED** [review]
This is definitely more of a refactor that I would have been comfortable doing! It looks really nice!

**Comment by inducer** [comment]
It occurred to me that it still equates "shape + order = polynomial space", which is not totally true. I'm thinking of introducing a separate object for the polynomial space. Is that crazy? (Not a big change, just `Pk` and `Qk` for now.)

**Comment by alexfikl** [comment]
> It occurred to me that it still equates "shape + order = polynomial space", which is not totally true. I'm thinking of introducing a separate object for the polynomial space. Is that crazy? (Not a big change, just `Pk` and `Qk` for now.)

Isn't that mostly what the discretization `ElementGroups` in meshmode are?

File: test/test_nodes.py
Comment (inducer) [reply test/test_nodes.py:40]: Just copied their example:

https://numpy.org/doc/stable/reference/random/generator.html

File: modepy/nodes.py
Comment (inducer) [reply modepy/nodes.py]: You mean `biunit_edge_clustered_...`? or `biunit_equispaced...`? Or drop one of them and only offer `biunit_` for edge-clustered?

**Comment by inducer** [comment]
> Isn't that mostly what the discretization `ElementGroups` in meshmode are?

In a way, sure... but they're about multiple elements, not just the reference element.

What worries me more is that we're currently answering the question "give me a basis for tetrahedra", which is nonsense. "Give me a basis for P^k" makes incrementally more sense. The motivating examples where things break down that I think about are pyramids and inhomogeneous tensor products.

File: modepy/nodes.py
Comment (alexfikl) [reply modepy/nodes.py]: I was thinking just `edge_clustered_nodes_for_shape` -> `biunit_nodes_for_shape`, mostly inspired by seeing `biunit_vertices_for_shape`.

File: test/test_nodes.py
Comment (alexfikl) [reply test/test_nodes.py:40]: TIL I thought the Mersenne twister was the default. Cool!

**Comment by alexfikl** [comment]
> What worries me more is that we're currently answering the question "give me a basis for tetrahedra", which is nonsense.

Yes, I agree it's a bit clunky. Especially as inducer/modepy#15 would already introduce a non-tensor product set for squares. It makes sense to add something like a polynomial space to express these things.

File: modepy/nodes.py
Comment (inducer) [reply modepy/nodes.py]: Not a fan. For `biunit_vertices_for_shape`, there is only one sensible answer (modulo order). For the nodes, there are many possible choices (technically, even "edge-clustered" doesn't *really* narrow it down). So I felt an adjective was warranted.

File: modepy/nodes.py
Comment (inducer) [reply modepy/nodes.py]: 299b32f

File: modepy/tools.py
Comment (inducer) [reply modepy/tools.py]: c75d799

File: modepy/modes.py
Comment (inducer) [reply modepy/modes.py:691]: 7f4e1a1 and 347027b

File: modepy/modes.py
Comment (inducer) [reply modepy/modes.py:713]: 7f4e1a1 and 347027b

File: modepy/shapes.py
Comment (inducer) [reply modepy/shapes.py]: c872de9a19cb34319be701d657060076e059d0c4

File: modepy/nodes.py
Comment (alexfikl) [reply modepy/nodes.py]: I guess I was thinking of it as having a function `biunit_nodes_for_shape` that just does a "give me some not horrible nodes of some sort", so it doesn't matter if it's a bit vague.

**Comment by alexfikl** [comment]
@inducer Should I merge this back into inducer/modepy#16 now?

**Comment by inducer** [comment]
Not yet. There's another big chunk coming.

File: modepy/nodes.py
Comment (alexfikl) [inline modepy/nodes.py]:
```suggestion
    return np.array(node_tuples_for_shape(shape, order))/order * 2 - 1
```
?

File: modepy/nodes.py
Comment (alexfikl) [inline modepy/nodes.py]:
```suggestion
    if shape.dim <= 3:
         return warp_and_blend_nodes(shape.dim, order)
    else:
         return equidistant_nodes(shape.dim, order)
```

File: modepy/nodes.py
Comment (alexfikl) [inline modepy/nodes.py]: Does this need to special case for 0-d anymore?

File: modepy/quadrature/__init__.py
Comment (alexfikl) [inline modepy/quadrature/__init__.py:5]:
```suggestion
.. autoclass:: TensorProductQuadrature
.. autoclass:: GaussLegendreTensorProductQuadrature
```
?

Should also generalize `TensorProductQuadrature` to non-homogeneous nD, to match `TensorProductBasis`.

File: modepy/quadrature/__init__.py
Comment (alexfikl) [inline modepy/quadrature/__init__.py]:
```suggestion
    if shape.dim == 0:
        quad = Quadrature(np.empty((0, 1)), np.empty((0, 1)))
    else:
        quad = LegendreGaussTensorProductQuadrature(order, shape.dim)
```

File: modepy/quadrature/__init__.py
Comment (alexfikl) [inline modepy/quadrature/__init__.py]:
```suggestion
    except (QuadratureRuleUnavailable, ValueError):
```

File: modepy/shapes.py
Comment (alexfikl) [inline modepy/shapes.py:335]: I sort of just copy pasted this from a test case. Does it match the hardcoded `face_vertex_indices` in meshmode?

**Review by alexfikl — COMMENTED** [review]
Added a few more comments. I can fix these after we merge back into inducer/modepy#16.

File: modepy/shapes.py
Comment (inducer) [reply modepy/shapes.py:335]: Those are going away. My plan is to make this one big apocalypse where meshmode gives up on knowing anything that modepy can tell it.

File: modepy/nodes.py
Comment (inducer) [reply modepy/nodes.py]: 7f5ec1e25f24cd074e53c78c3a6da117898c62ca

**Comment by inducer** [comment]
The "big chunk of stuff" is now in, mostly consisting of introducing a notion of polynomial spaces. Take a look. Good to merge back once the checks on https://github.com/inducer/meshmode/pull/95 pass.

File: modepy/quadrature/__init__.py
Comment (alexfikl) [inline modepy/quadrature/__init__.py:151]: If only Python had multiple dispatch like Julia :cry:

File: modepy/spaces.py
Comment (alexfikl) [inline modepy/spaces.py:68]: Just `dim`? It feels a bit redundant to write `space.space_dim`.

**Review by alexfikl — COMMENTED** [review]
It looks good to me! inducer/meshmode#95 also seems to have passed too!

I'll merge this back into inducer/modepy#16 and we can nitpick it more there!

File: modepy/quadrature/__init__.py
Comment (inducer) [reply modepy/quadrature/__init__.py:151]: Agreed!

In fact, I briefly played with https://pypi.org/project/multipledispatch/ for this, but there were a number of (fixable!) shortcomings there that made me not like that particular realization. (around the `Dispatcher` and how they handle namespacing) Basically, if there were a multiple-dispatch package that works like `singledispatch`, I'd use it in a heartbeat.

At any rate, I figured we can easily incorporate that once a suitable solution shows up, starting from this interface.

File: modepy/nodes.py
Comment (alexfikl) [inline modepy/nodes.py:452]: Should this not be the same as `tensor_product_nodes(dim, equidistant_nodes(1, order))`?

I guess the problem is that there's now an inconsistency between the `node_tuples` and `tensor_product_nodes` order, since the nodes got flipped.

File: modepy/spaces.py
Comment (inducer) [reply modepy/spaces.py:68]: TBH, I hate both of these names. I picked them to be obnoxiously similar because I wanted to sow enough confusion that people would stumble and realize the difference. `dim` IMO is bad because it could easily/justifiably be either.

But I'm open to other suggestions.

Comment (alexfikl) [reply modepy/spaces.py:68]: Yeah, I think I just convinced myself that it makes more sense to call `spatial_dim` just `dim`, since it's used like that everywhere else in the code.

Can't really think of anything nicer either, I'm afraid.

File: modepy/modes.py
Comment (alexfikl) [inline modepy/modes.py:836]: Should this also take a shape as an argument? I'm thinking that something like inducer/modepy#15 needs a PN basis on the square.

Comment (inducer) [reply modepy/modes.py:836]: I think you're right. At first, I thought, "A P^N basis is a P^N basis", which is true, but its Vdm conditioning is going to depend on geometry.

:+1: for passing a shape here, too.

File: modepy/nodes.py
Comment (inducer) [reply modepy/nodes.py:452]: Out of curiosity: how'd this get resolved?

File: modepy/modes.py
Comment (alexfikl) [reply modepy/modes.py:836]: Added in 97abd45

Comment (inducer) [reply modepy/modes.py:836]: Thanks, LGTM!

**Pull request merged**
